heka/sandbox/decoders/extract_telemetry_dimensions.lua

```diff
@@ -93,120 +106,117 @@ function parse_creation_date(date)
-        msg.Fields.docType          = parsed.type or UNK_DIM
+        update_field(msg.Fields, "sourceVersion", tostring(parsed.version))
+        update_field(msg.Fields, "docType"      , parsed.type or UNK_DIM)
 
         -- Get some more dimensions.
-        msg.Fields.appName          = app.name or UNK_DIM
-        msg.Fields.appVersion       = app.version or UNK_DIM
-        msg.Fields.appUpdateChannel = app.channel or UNK_DIM
+        update_field(msg.Fields, "appName"           , app.name or UNK_DIM)
+        update_field(msg.Fields, "appVersion"        , app.version or UNK_DIM)
+        update_field(msg.Fields, "appUpdateChannel"  , app.channel or UNK_DIM)
 
         -- Do not want default values for these.
-        msg.Fields.appBuildId       = app.buildId
-        msg.Fields.appVendor        = app.vendor
-        msg.Fields.clientId         = parsed.clientId
+        update_field(msg.Fields, "appBuildId", app.buildId)
+        update_field(msg.Fields, "appVendor" , app.vendor)
+        clientId = parsed.clientId
+        update_field(msg.Fields, "clientId"  , clientId)
 
         -- restructure the main ping message
         if parsed.type == "main" then
-            split_objects(parsed.environment, "environment", environment_objects)
-            split_objects(parsed.payload, "payload", main_ping_objects)
+            split_objects(msg.Fields, parsed.environment, "environment", environment_objects)
+            split_objects(msg.Fields, parsed.payload, "payload", main_ping_objects)
             local ok, json = pcall(cjson.encode, parsed) -- re-encode the remaining data
-            if not ok then return -1, json end
+            if not ok then return json end
             msg.Payload = json
         else
             msg.Payload = json
         end
     end
+    update_field(msg.Fields, "sampleId", sample(clientId, 100))
+end
 
-    -- Carry forward more incoming fields.
-    msg.Fields.geoCountry = read_message("Fields[geoCountry]") or UNK_GEO
-    msg.Timestamp         = read_message("Timestamp")
-    msg.Fields.Host       = read_message("Fields[Host]")
-    msg.Fields.DNT        = read_message("Fields[DNT]")
-    msg.Fields.clientDate = read_message("Fields[Date]")
-
-    msg.Fields.submissionDate = os.date("%Y%m%d", msg.Timestamp / 1e9)
-
-    msg.Fields.sampleId = sample(msg.Fields.clientId, 100)
-
-    -- Send new message along.
-    local err
-    ok, err = pcall(inject_message, msg)
+local function send_message(msg, phase, err)
+    if err then
+        msg.Type = "telemetry.error"
+        update_field(msg.Fields, "DecodeErrorType", phase)
+        update_field(msg.Fields, "DecodeError", err)
+    end
+    local ok, err = pcall(inject_message, msg)
     if not ok then
         return -1, err
     end
     return 0
 end
+
+function process_message()
+    local raw = read_message("raw")
+    local ok, msg = pcall(decode_message, raw)
+    if not ok then return -1, msg end
```

Comment:
There is nothing to forward the Heka message is crap, so it is logged as an internal ProcessMessageFailure.